io-engine/src/rebuild/rebuild_log.rs

```diff
@@ -0,0 +1,135 @@
+use bit_vec::BitVec;
+use parking_lot::Mutex;
+use spdk_rs::IoType;
+use std::{
+    fmt::{Debug, Formatter},
+    sync::Arc,
+};
+
+use super::SEGMENT_SIZE;
+
+// Returns ceil of an integer division.
+fn div_ceil(a: u64, b: u64) -> u64 {
+    (a + b - 1) / b
+}
+
+/// Rebuild log data structure.
+pub struct RebuildLog {
+    /// Name of the underlying device, stored for diagnostic purposes.
+    device_name: String,
+    /// Segment bitmap. Ones indicate modified segments.
+    segments: BitVec,
+    /// Device size in segments.
+    num_segments: u64,
+    /// Device size in block.
+    num_blocks: u64,
+    /// Size of block in bytes.
+    block_len: u64,
+}
+
+impl Debug for RebuildLog {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Rebuild log: '{}' ({} segments: {} blocks x {})",
+            self.device_name,
+            self.num_segments,
+            self.num_blocks,
+            self.block_len
+        )
+    }
+}
+
+impl RebuildLog {
+    /// Creates a new rebuild log instance for the given device.
+    pub fn new(device_name: &str, num_blocks: u64, block_len: u64) -> Self {
+        let num_segments = div_ceil(num_blocks * block_len, SEGMENT_SIZE);
+        let mut segments = BitVec::new();
+        segments.grow(num_segments as usize, false);
+        Self {
+            device_name: device_name.to_owned(),
+            segments,
+            num_segments,
+            num_blocks,
+            block_len,
+        }
+    }
+
+    /// Logs the given operation, marking the corresponding segment as modified
+    /// in the case of a write operation.
+    fn log_op(&mut self, _io_type: IoType, block_offset: u64, num_blocks: u64) {
+        let start_seg = self.segment_index(block_offset);
+        let end_seg = self.segment_index(block_offset + num_blocks);
+        for i in start_seg ..= end_seg {
+            self.segments.set(i, true);
+        }
+    }
+
+    /// Determines if the given block was modified by a write operation.
+    fn is_modified(&self, block_offset: u64) -> bool {
+        let seg = self.segment_index(block_offset);
+        if seg < self.segments.len() {
+            self.segments[seg]
+        } else {
+            true
+        }
+    }
+
+    /// Calculates the index of segment corresponding to a given block.
+    fn segment_index(&self, block_offset: u64) -> usize {
+        (block_offset * self.block_len / SEGMENT_SIZE) as usize
+    }
+
+    /// Counts the total number of modified blocks.
+    pub fn count_modified_blocks(&self) -> u64 {
+        let segs = self.segments.iter().filter(|i| *i).count() as u64;
+        segs * SEGMENT_SIZE / self.block_len
+    }
+}
+
+/// A handle object for a rebuild log instance.
+/// A rebuild log instance allows to have multiple handle to access it.
```

Comment:
This is part of RUST learning. Can you elaborate on this? Why do we need multiple instances? Even if it needs multiple instances can't that be handled without having this wrapper?